Modified WSRequest.withRequestTimeout to take Duration as a parameter

njrich28 opened the pull request:
Fixes #4846

As per discussion on issue, I've modified the API to take a `scala.concurrent.duration.Duration` as the parameter rather than `Long`.

I've modified the Ning implementation to set the requestTimeout to -1 if `Duration.Inf` is passed and to validate all non-infinite values to make sure they are non-negative and less than `Int.MaxValue`. Tests and documentation updated.

marcospereira commented:
This should be fixed for the Java API too. Currently it is throwing an exception when timeout is less then 0.

njrich28 commented:
@marcospereira happy to fix the Java API too - any suggestions for what to use instead of `scala.concurrent.duration.Duration`? I originally looked at `java.time.Duration` but this doesn't have an obvious infinite representation.

jroper commented:
The Java API can just support -1 - Java is inherently less type safe than Scala, no point fighting it :)

njrich28 commented:
I've fixed the Java API to support a request timeout of -1 and added tests and documentation. Also amended Scaladoc to qualify warning about interruption of stream consumption.

marcospereira commented on framework/src/play-java-ws/src/main/java/play/libs/ws/ning/NingWSRequest.java:
```diff
@@ -175,8 +175,8 @@ public WSRequest setVirtualHost(String virtualHost) {
-        if (timeout < 0 || timeout > Integer.MAX_VALUE) {
-            throw new IllegalArgumentException("Timeout must be between 0 and " + Integer.MAX_VALUE + " inclusive");
+        if (timeout < -1 || timeout > Integer.MAX_VALUE) {
+            throw new IllegalArgumentException("Timeout must be between -1 and " + Integer.MAX_VALUE + " inclusive");
```
Maybe you can change this message to: `"Timeout must be between -1(which represents infinity here) and " + Integer.MAX_VALUE + " inclusive""`.

marcospereira commented:
I've made a very small suggestion. Besides that, LGTM.

jroper commented:
LGTM.
